File: backend/app/schemas/knowledge_base.py
"""
Knowledge Base Schemas - Refactored
"""
from pydantic import BaseModel
from typing import Optional, List, Any, Dict
from datetime import datetime
from enum import Enum


class DocumentType(str, Enum):
    PITCH_DECK = "pitch_deck"
    CASE_STUDY = "case_study"
    PRICING = "pricing"
    PRODUCT_INFO = "product_info"
    COMPETITOR_ANALYSIS = "competitor_analysis"
    EMAIL_TEMPLATES = "email_templates"
    OTHER = "other"


# ============ Document Folder Schemas ============

class DocumentFolderBase(BaseModel):
    name: str
    parent_id: Optional[int] = None


class DocumentFolderCreate(DocumentFolderBase):
    pass


class DocumentFolderResponse(DocumentFolderBase):
    id: int
    created_at: datetime
    
    class Config:
        from_attributes = True


# ============ Document Schemas ============

class DocumentBase(BaseModel):
    name: str
    document_type: DocumentType = DocumentType.OTHER
    folder_id: Optional[int] = None


class DocumentCreate(DocumentBase):
    pass


class DocumentUpdate(BaseModel):
    name: Optional[str] = None
    document_type: Optional[DocumentType] = None
    folder_id: Optional[int] = None


class DocumentResponse(DocumentBase):
    id: int
    original_filename: Optional[str]
    content_md: Optional[str]
    status: str
    error_message: Optional[str]
    created_at: datetime
    updated_at: Optional[datetime]
    
    class Config:
        from_attributes = True


# ============ Company Profile Schemas (Simplified) ============

class CompanyProfileBase(BaseModel):
    name: Optional[str] = None
    website: Optional[str] = None
    summary: Optional[str] = None


class CompanyProfileCreate(CompanyProfileBase):
    pass


class CompanyProfileUpdate(CompanyProfileBase):
    pass


class CompanyProfileResponse(CompanyProfileBase):
    id: int
    created_at: datetime
    updated_at: Optional[datetime]
    
    class Config:
        from_attributes = True


# ============ Product Schemas ============

class ProductBase(BaseModel):
    name: str
    description: Optional[str] = None
    features: Optional[List[str]] = None
    pricing: Optional[Dict[str, Any]] = None
    target_segment_ids: Optional[List[int]] = None
    email_snippet: Optional[str] = None
    is_active: bool = True
    sort_order: int = 0


class ProductCreate(ProductBase):
    pass


class ProductUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    features: Optional[List[str]] = None
    pricing: Optional[Dict[str, Any]] = None
    target_segment_ids: Optional[List[int]] = None
    email_snippet: Optional[str] = None
    is_active: Optional[bool] = None
    sort_order: Optional[int] = None


class ProductResponse(ProductBase):
    id: int
    created_at: datetime
    updated_at: Optional[datetime]
    
    class Config:
        from_attributes = True


# ============ Segment Column Schemas ============

class SegmentColumnBase(BaseModel):
    name: str
    display_name: str
    column_type: str = "text"  # text, number, list, rich_text, case_select
    is_required: bool = False
    sort_order: int = 0
    options: Optional[List[str]] = None


class SegmentColumnCreate(SegmentColumnBase):
    pass


class SegmentColumnUpdate(BaseModel):
    display_name: Optional[str] = None
    column_type: Optional[str] = None
    is_required: Optional[bool] = None
    sort_order: Optional[int] = None
    options: Optional[List[str]] = None


class SegmentColumnResponse(SegmentColumnBase):
    id: int
    is_system: bool
    created_at: datetime
    
    class Config:
        from_attributes = True


# ============ Segment Schemas ============

class SegmentBase(BaseModel):
    name: str
    data: Dict[str, Any] = {}
    is_active: bool = True
    sort_order: int = 0


class SegmentCreate(SegmentBase):
    pass


class SegmentUpdate(BaseModel):
    name: Optional[str] = None
    data: Optional[Dict[str, Any]] = None
    is_active: Optional[bool] = None
    sort_order: Optional[int] = None


class SegmentResponse(SegmentBase):
    id: int
    created_at: datetime
    updated_at: Optional[datetime]
    
    class Config:
        from_attributes = True


# ============ Competitor Schemas ============

class CompetitorBase(BaseModel):
    name: str
    website: Optional[str] = None
    description: Optional[str] = None
    their_strengths: Optional[List[str]] = None
    their_weaknesses: Optional[List[str]] = None
    our_advantages: Optional[List[str]] = None
    their_positioning: Optional[str] = None
    price_comparison: Optional[str] = None
    notes: Optional[str] = None


class CompetitorCreate(CompetitorBase):
    pass


class CompetitorUpdate(BaseModel):
    name: Optional[str] = None
    website: Optional[str] = None
    description: Optional[str] = None
    their_strengths: Optional[List[str]] = None
    their_weaknesses: Optional[List[str]] = None
    our_advantages: Optional[List[str]] = None
    their_positioning: Optional[str] = None
    price_comparison: Optional[str] = None
    notes: Optional[str] = None


class CompetitorResponse(CompetitorBase):
    id: int
    created_at: datetime
    updated_at: Optional[datetime]
    
    class Config:
        from_attributes = True


# ============ Case Study Schemas ============

class CaseStudyBase(BaseModel):
    client_name: str
    client_website: Optional[str] = None
    client_industry: Optional[str] = None
    client_size: Optional[str] = None
    challenge: Optional[str] = None
    solution: Optional[str] = None
    results: Optional[str] = None
    key_metrics: Optional[Dict[str, str]] = None
    testimonial: Optional[str] = None
    testimonial_author: Optional[str] = None
    testimonial_title: Optional[str] = None
    email_snippet: Optional[str] = None
    is_public: bool = True


class CaseStudyCreate(CaseStudyBase):
    pass


class CaseStudyUpdate(BaseModel):
    client_name: Optional[str] = None
    client_website: Optional[str] = None
    client_industry: Optional[str] = None
    client_size: Optional[str] = None
    challenge: Optional[str] = None
    solution: Optional[str] = None
    results: Optional[str] = None
    key_metrics: Optional[Dict[str, str]] = None
    testimonial: Optional[str] = None
    testimonial_author: Optional[str] = None
    testimonial_title: Optional[str] = None
    email_snippet: Optional[str] = None
    is_public: Optional[bool] = None


class CaseStudyResponse(CaseStudyBase):
    id: int
    created_at: datetime
    updated_at: Optional[datetime]
    
    class Config:
        from_attributes = True


# ============ Voice Tone Schemas ============

class VoiceToneBase(BaseModel):
    name: str
    description: Optional[str] = None
    personality_traits: Optional[List[str]] = None
    writing_style: Optional[str] = None
    do_use: Optional[List[str]] = None
    dont_use: Optional[List[str]] = None
    example_messages: Optional[List[str]] = None
    formality_level: int = 5
    emoji_usage: bool = False
    is_default: bool = False


class VoiceToneCreate(VoiceToneBase):
    pass


class VoiceToneUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    personality_traits: Optional[List[str]] = None
    writing_style: Optional[str] = None
    do_use: Optional[List[str]] = None
    dont_use: Optional[List[str]] = None
    example_messages: Optional[List[str]] = None
    formality_level: Optional[int] = None
    emoji_usage: Optional[bool] = None
    is_default: Optional[bool] = None


class VoiceToneResponse(VoiceToneBase):
    id: int
    created_at: datetime
    updated_at: Optional[datetime]
    
    class Config:
        from_attributes = True


# ============ Booking Link Schemas (Simplified) ============

class BookingLinkBase(BaseModel):
    name: str
    url: str
    when_to_use: Optional[str] = None
    is_active: bool = True


class BookingLinkCreate(BookingLinkBase):
    pass


class BookingLinkUpdate(BaseModel):
    name: Optional[str] = None
    url: Optional[str] = None
    when_to_use: Optional[str] = None
    is_active: Optional[bool] = None


class BookingLinkResponse(BookingLinkBase):
    id: int
    created_at: datetime
    updated_at: Optional[datetime]
    
    class Config:
        from_attributes = True


# ============ Blocklist Schemas (Extended) ============

class BlocklistBase(BaseModel):
    domain: Optional[str] = None
    email: Optional[str] = None
    company_name: Optional[str] = None
    reason: Optional[str] = None


class BlocklistCreate(BlocklistBase):
    pass


class BlocklistResponse(BlocklistBase):
    id: int
    created_at: datetime
    
    class Config:
        from_attributes = True


# ============ Import/Export Schemas ============

class CSVImportResult(BaseModel):
    success: bool
    imported_count: int
    errors: List[str] = []


class AIImportRequest(BaseModel):
    text: str
    entity_type: str
    save_to_db: bool = False
    parse_multiple: bool = False


class AIImportResponse(BaseModel):
    success: bool
    entity_type: str
    data: Optional[Dict[str, Any]] = None
    saved_ids: Optional[List[int]] = None
    error: Optional[str] = None
    tokens_used: int = 0


# ============ Export Schema ============

class KnowledgeBaseExport(BaseModel):
    company: Optional[CompanyProfileResponse] = None
    products: List[ProductResponse] = []
    segments: List[SegmentResponse] = []
    segment_columns: List[SegmentColumnResponse] = []
    competitors: List[CompetitorResponse] = []
    case_studies: List[CaseStudyResponse] = []
    voice_tones: List[VoiceToneResponse] = []
    booking_links: List[BookingLinkResponse] = []
    blocklist: List[BlocklistResponse] = []
    documents: List[DocumentResponse] = []

File: frontend/src/api/knowledgeBase.ts
import api from './client';

const BASE = '/knowledge-base';

// ============ Document Folders ============

export interface DocumentFolder {
  id: number;
  name: string;
  parent_id?: number;
  created_at: string;
}

export const getFolders = async (): Promise<DocumentFolder[]> => {
  const response = await api.get(`${BASE}/folders`);
  return response.data;
};

export const createFolder = async (data: { name: string; parent_id?: number }): Promise<DocumentFolder> => {
  const response = await api.post(`${BASE}/folders`, data);
  return response.data;
};

export const deleteFolder = async (id: number): Promise<void> => {
  await api.delete(`${BASE}/folders/${id}`);
};

// ============ Documents ============

export interface KBDocument {
  id: number;
  name: string;
  original_filename?: string;
  document_type: string;
  folder_id?: number;
  content_md?: string;
  status: string;
  error_message?: string;
  created_at: string;
  updated_at?: string;
}

export const getDocuments = async (folderId?: number): Promise<KBDocument[]> => {
  const params = folderId !== undefined ? { folder_id: folderId } : {};
  const response = await api.get(`${BASE}/documents`, { params });
  return response.data;
};

export const uploadDocument = async (
  file: File,
  name?: string,
  documentType?: string,
  folderId?: number
): Promise<KBDocument> => {
  const formData = new FormData();
  formData.append('file', file);
  if (name) formData.append('name', name);
  if (documentType) formData.append('document_type', documentType);
  if (folderId) formData.append('folder_id', String(folderId));
  
  const response = await api.post(`${BASE}/documents/upload`, formData, {
    headers: { 'Content-Type': 'multipart/form-data' }
  });
  return response.data;
};

export const updateDocument = async (id: number, data: Partial<KBDocument>): Promise<KBDocument> => {
  const response = await api.patch(`${BASE}/documents/${id}`, data);
  return response.data;
};

export const deleteDocument = async (id: number): Promise<void> => {
  await api.delete(`${BASE}/documents/${id}`);
};

export const regenerateSummary = async (): Promise<{ summary: string; document_count: number }> => {
  const response = await api.post(`${BASE}/documents/regenerate-summary`);
  return response.data;
};

// ============ Company Profile ============

export interface CompanyProfile {
  id: number;
  name?: string;
  website?: string;
  summary?: string;
  created_at: string;
  updated_at?: string;
}

export const getCompanyProfile = async (): Promise<CompanyProfile | null> => {
  const response = await api.get(`${BASE}/company`);
  return response.data;
};

export const saveCompanyProfile = async (data: Partial<CompanyProfile>): Promise<CompanyProfile> => {
  const response = await api.put(`${BASE}/company`, data);
  return response.data;
};

// ============ Products ============

export interface Product {
  id: number;
  name: string;
  description?: string;
  features?: string[];
  pricing?: Record<string, any>;
  target_segment_ids?: number[];
  email_snippet?: string;
  is_active: boolean;
  sort_order: number;
  created_at: string;
  updated_at?: string;
}

export const getProducts = async (): Promise<Product[]> => {
  const response = await api.get(`${BASE}/products`);
  return response.data;
};

export const createProduct = async (data: Partial<Product>): Promise<Product> => {
  const response = await api.post(`${BASE}/products`, data);
  return response.data;
};

export const updateProduct = async (id: number, data: Partial<Product>): Promise<Product> => {
  const response = await api.patch(`${BASE}/products/${id}`, data);
  return response.data;
};

export const deleteProduct = async (id: number): Promise<void> => {
  await api.delete(`${BASE}/products/${id}`);
};

// ============ Segment Columns ============

export interface SegmentColumn {
  id: number;
  name: string;
  display_name: string;
  column_type: 'text' | 'number' | 'list' | 'rich_text' | 'case_select';
  is_system: boolean;
  is_required: boolean;
  sort_order: number;
  options?: string[];
  created_at: string;
}

export const getSegmentColumns = async (): Promise<SegmentColumn[]> => {
  const response = await api.get(`${BASE}/segment-columns`);
  return response.data;
};

export const createSegmentColumn = async (data: Partial<SegmentColumn>): Promise<SegmentColumn> => {
  const response = await api.post(`${BASE}/segment-columns`, data);
  return response.data;
};

export const updateSegmentColumn = async (id: number, data: Partial<SegmentColumn>): Promise<SegmentColumn> => {
  const response = await api.patch(`${BASE}/segment-columns/${id}`, data);
  return response.data;
};

export const deleteSegmentColumn = async (id: number): Promise<void> => {
  await api.delete(`${BASE}/segment-columns/${id}`);
};

// ============ Segments ============

export interface Segment {
  id: number;
  name: string;
  data: Record<string, any>;
  is_active: boolean;
  sort_order: number;
  created_at: string;
  updated_at?: string;
}

export const getSegments = async (): Promise<Segment[]> => {
  const response = await api.get(`${BASE}/segments`);
  return response.data;
};

export const createSegment = async (data: Partial<Segment>): Promise<Segment> => {
  const response = await api.post(`${BASE}/segments`, data);
  return response.data;
};

export const updateSegment = async (id: number, data: Partial<Segment>): Promise<Segment> => {
  const response = await api.patch(`${BASE}/segments/${id}`, data);
  return response.data;
};

export const deleteSegment = async (id: number): Promise<void> => {
  await api.delete(`${BASE}/segments/${id}`);
};

export const importSegmentsCSV = async (file: File): Promise<CSVImportResult> => {
  const formData = new FormData();
  formData.append('file', file);
  const response = await api.post(`${BASE}/segments/import-csv`, formData, {
    headers: { 'Content-Type': 'multipart/form-data' }
  });
  return response.data;
};

// ============ Competitors ============

export interface Competitor {
  id: number;
  name: string;
  website?: string;
  description?: string;
  their_strengths?: string[];
  their_weaknesses?: string[];
  our_advantages?: string[];
  their_positioning?: string;
  price_comparison?: string;
  notes?: string;
  created_at: string;
  updated_at?: string;
}

export const getCompetitors = async (): Promise<Competitor[]> => {
  const response = await api.get(`${BASE}/competitors`);
  return response.data;
};

export const createCompetitor = async (data: Partial<Competitor>): Promise<Competitor> => {
  const response = await api.post(`${BASE}/competitors`, data);
  return response.data;
};

export const updateCompetitor = async (id: number, data: Partial<Competitor>): Promise<Competitor> => {
  const response = await api.patch(`${BASE}/competitors/${id}`, data);
  return response.data;
};

export const deleteCompetitor = async (id: number): Promise<void> => {
  await api.delete(`${BASE}/competitors/${id}`);
};

export const importCompetitorsCSV = async (file: File): Promise<CSVImportResult> => {
  const formData = new FormData();
  formData.append('file', file);
  const response = await api.post(`${BASE}/competitors/import-csv`, formData, {
    headers: { 'Content-Type': 'multipart/form-data' }
  });
  return response.data;
};

// ============ Case Studies ============

export interface CaseStudy {
  id: number;
  client_name: string;
  client_website?: string;
  client_industry?: string;
  client_size?: string;
  challenge?: string;
  solution?: string;
  results?: string;
  key_metrics?: Record<string, string>;
  testimonial?: string;
  testimonial_author?: string;
  testimonial_title?: string;
  email_snippet?: string;
  is_public: boolean;
  created_at: string;
  updated_at?: string;
}

export const getCaseStudies = async (): Promise<CaseStudy[]> => {
  const response = await api.get(`${BASE}/case-studies`);
  return response.data;
};

export const createCaseStudy = async (data: Partial<CaseStudy>): Promise<CaseStudy> => {
  const response = await api.post(`${BASE}/case-studies`, data);
  return response.data;
};

export const updateCaseStudy = async (id: number, data: Partial<CaseStudy>): Promise<CaseStudy> => {
  const response = await api.patch(`${BASE}/case-studies/${id}`, data);
  return response.data;
};

export const deleteCaseStudy = async (id: number): Promise<void> => {
  await api.delete(`${BASE}/case-studies/${id}`);
};

// ============ Voice Tones ============

export interface VoiceTone {
  id: number;
  name: string;
  description?: string;
  personality_traits?: string[];
  writing_style?: string;
  do_use?: string[];
  dont_use?: string[];
  example_messages?: string[];
  formality_level: number;
  emoji_usage: boolean;
  is_default: boolean;
  created_at: string;
  updated_at?: string;
}

export const getVoiceTones = async (): Promise<VoiceTone[]> => {
  const response = await api.get(`${BASE}/voice-tones`);
  return response.data;
};

export const createVoiceTone = async (data: Partial<VoiceTone>): Promise<VoiceTone> => {
  const response = await api.post(`${BASE}/voice-tones`, data);
  return response.data;
};

export const updateVoiceTone = async (id: number, data: Partial<VoiceTone>): Promise<VoiceTone> => {
  const response = await api.patch(`${BASE}/voice-tones/${id}`, data);
  return response.data;
};

export const deleteVoiceTone = async (id: number): Promise<void> => {
  await api.delete(`${BASE}/voice-tones/${id}`);
};

// ============ Booking Links ============

export interface BookingLink {
  id: number;
  name: string;
  url: string;
  when_to_use?: string;
  is_active: boolean;
  created_at: string;
  updated_at?: string;
}

export const getBookingLinks = async (): Promise<BookingLink[]> => {
  const response = await api.get(`${BASE}/booking-links`);
  return response.data;
};

export const createBookingLink = async (data: Partial<BookingLink>): Promise<BookingLink> => {
  const response = await api.post(`${BASE}/booking-links`, data);
  return response.data;
};

export const updateBookingLink = async (id: number, data: Partial<BookingLink>): Promise<BookingLink> => {
  const response = await api.patch(`${BASE}/booking-links/${id}`, data);
  return response.data;
};

export const deleteBookingLink = async (id: number): Promise<void> => {
  await api.delete(`${BASE}/booking-links/${id}`);
};

// ============ Blocklist ============

export interface BlocklistEntry {
  id: number;
  domain?: string;
  email?: string;
  company_name?: string;
  reason?: string;
  created_at: string;
}

export const getBlocklist = async (): Promise<BlocklistEntry[]> => {
  const response = await api.get(`${BASE}/blocklist`);
  return response.data;
};

export const addToBlocklist = async (data: Partial<BlocklistEntry>): Promise<BlocklistEntry> => {
  const response = await api.post(`${BASE}/blocklist`, data);
  return response.data;
};

export const removeFromBlocklist = async (id: number): Promise<void> => {
  await api.delete(`${BASE}/blocklist/${id}`);
};

export const importBlocklistCSV = async (file: File): Promise<CSVImportResult> => {
  const formData = new FormData();
  formData.append('file', file);
  const response = await api.post(`${BASE}/blocklist/import-csv`, formData, {
    headers: { 'Content-Type': 'multipart/form-data' }
  });
  return response.data;
};

// ============ Tags & Context ============

export interface KBTag {
  tag: string;
  label: string;
  description: string;
  type: 'summary' | 'company' | 'product' | 'segment' | 'voice' | 'case' | 'booking' | 'competitor' | 'document';
}

export const getAvailableTags = async (): Promise<{ tags: KBTag[] }> => {
  const response = await api.get(`${BASE}/tags`);
  return response.data;
};

// ============ Import ============

export interface CSVImportResult {
  success: boolean;
  imported_count: number;
  errors: string[];
}

export interface AIImportRequest {
  text: string;
  entity_type: string;
  save_to_db?: boolean;
  parse_multiple?: boolean;
}

export interface AIImportResponse {
  success: boolean;
  entity_type: string;
  data?: Record<string, any>;
  saved_ids?: number[];
  error?: string;
  tokens_used: number;
}

export const aiImport = async (request: AIImportRequest): Promise<AIImportResponse> => {
  const response = await api.post(`${BASE}/ai-import`, request);
  return response.data;
};

// ============ Export ============

export const exportKnowledgeBase = async () => {
  const response = await api.get(`${BASE}/export`);
  return response.data;
};
